Initial FreeIPA Integration

abstractj opened the pull request:
- Provide username/password authentication with PAM
- Obtain user data from SSSD
- Feature packs for dbus-java, libpam4j and SSSD API
- Provisioning script

Note: [The test suite was not included yet](https://github.com/abstractj/keycloak/pull/1). I would like to review and create a separate pull-request for it. If it's ok.
## Testing instructions

These testing instructions are pretty much based on this [README](https://github.com/abstractj/keycloak/blob/KEYCLOAK-3036/federation/freeipa/README.md)
1. Start a FreeIPA server, you can do that my running the docker image available
   
   ```
   docker run --name freeipa-server-container -it \
   -h server.arsenic.local -e PASSWORD=password \
   -v /sys/fs/cgroup:/sys/fs/cgroup:ro \
   -v /var/lib/ipa-data:/data:Z adelton/freeipa-server
   ```
   
   Where `server.arsenic.local` is the hostname desired. 
2. After the docker image initialization, get the IP address and make sure to add it to /etc/hosts
   
   ```
   172.17.0.2     server.arsenic.local
   ```
3. Enroll a client machine to the FreeIPA server
   
   ```
   ipa-client-install --mkhomedir -p admin -w password --force --force-join
   ```
   
   The reason why `--force` and `--force-join` were used, it's because I faced some DNS issues with IPA and Docker. But in theory, it should work without these options.
4. Check if it worked running:
   
   ```
   kinit admin
   ```
   
   You should be prompted for the password.
5. Add some user for testing purposes
   
   ```
   $ ipa user-add john --first=John --last=Smith --email=john@smith.com --phone=042424242 --street="Testing street" \      --city="Testing city" --state="Testing State" --postalcode=0000000000
   ```
6. Download and install the RPM package for DBus Java
   
   ```
   $ wget -c https://github.com/abstractj/libunix-dbus-java/releases/download/libunix-dbus-java-0.8.0/libunix-dbus-java-0.8.0-1.fc24.x86_64.rpm
   ```
7. Run the provisioning script
   
   ```
   $KEYCLOAK_HOME/bin/federation-sssd-setup.sh 
   ```
   
   This will do the proper changes to `/etc/sssd/sssd.conf` and add `keycloak` file to `/etc/pam.d/`. After running this script, your `sssd.conf` should look like [this](https://github.com/abstractj/keycloak/blob/KEYCLOAK-3036/federation/freeipa/README.md#sssdconf).
8. Only to make sure that everything is working as expected, run dbus-send
   
   ```
   sudo dbus-send --print-reply --system --dest=org.freedesktop.sssd.infopipe /org/freedesktop/sssd/infopipe org.freedesktop.sssd.infopipe.GetUserGroups string:john
   ```
   
   If this command do not return user's group, it means that SSSDFederation provider won't be able to do that on Keycloak too. Most of the time it happens because the machine is not properly enrolled to FreeIPA server or you don't have the rights for it. 

If you don't have the rights, check your sssd.conf. The user running Keycloak, must be included in this section:

```
  [ifp]
  allowed_uids = root
```

If everything went well, start Keycloak and enable SSSDFederation provider. Try to authenticate as John.

Commit:
Initial FreeIPA Integration

  - Provide username/password authentication with PAM
  - Obtain user data from SSSD
  - Feature packs for dbus-java, libpam4j and SSSD API
  - Provisioning script

stianst commented:
There's a failing test:
UserFederationTest.testProviderFactories

stianst commented:
I would like this to be refactored to a single module (federation/freeipa) and not have separate modules for dbus/pam/sssd. We don't want to advertise general usage of those libraries so they should be hidden within the freeipa module. Further, we also do want to reduce the amount of modules in Keycloak (we just did a huge job in the past to do this).

Please remove all README files. Documentation should only exist in Gitbook and duplicated in Github repository.

For keycloak-dbus-java module I'd remove changelog and debug.conf. The AUTHORS and COPYING I'm less sure about, but they should probably just be moved to src/resources and renamed to DBUS-JAVA-AUTHORS/LICENSE. By the way did we check that including these libraries is fine from legal perspective? We probably need to confirm this.

Remove things like https://github.com/abstractj/keycloak/blob/KEYCLOAK-3036/federation/freeipa/keycloak-libpam4j/src/site/site.xml as well please.

abstractj commented:
@stianst I believe I did all the changes you mentioned. Just some notes:

* Instead of rename it to federation/freeipa, I believe it makes more sense to be federation/sssd, that's why I moved all the code to keycloak-sssd-federation
* The license files, for clarity I renamed it under the prefix DBUS-JAVA
